clang/runtime/dpct-rt/include/dpl_extras/algorithm.h.inc

```diff
@@ -1160,6 +1160,277 @@ partition(Policy &&policy, Iter1 first, Iter1 last, Iter2 mask, Pred p) {
+                bool descending, int begin_bit, int end_bit) {
+  using key_t_value_t = typename ::std::iterator_traits<key_t>::value_type;
+
+  int clipped_begin_bit = ::std::max(begin_bit, 0);
+  int clipped_end_bit =
+      ::std::min((::std::uint64_t)end_bit, sizeof(key_t_value_t) * 8);
+  int num_bytes = (clipped_end_bit - clipped_begin_bit - 1) / 8 + 1;
+
+  auto transform_and_sort_pairs_f = [&](auto x) {
+    using T = typename ::std::decay_t<decltype(x)>;
+    internal::transform_and_sort_pairs<decltype(policy), key_t, key_out_t, T,
+                                       value_t, value_out_t>(
+        ::std::forward<_ExecutionPolicy>(policy), keys_in, keys_out, values_in,
+        values_out, n, descending, clipped_begin_bit, clipped_end_bit);
+  };
+
+  if (clipped_end_bit - clipped_begin_bit == sizeof(key_t_value_t) * 8) {
+    internal::sort_only_pairs(::std::forward<_ExecutionPolicy>(policy), keys_in,
+                              keys_out, values_in, values_out, n, descending);
+  } else if (num_bytes == 1) {
+    transform_and_sort_pairs_f.template operator()<uint8_t>(0);
+  } else if (num_bytes == 2) {
+    transform_and_sort_pairs_f.template operator()<uint16_t>(0);
+  } else if (num_bytes <= 4) {
+    transform_and_sort_pairs_f.template operator()<uint32_t>(0);
+  } else // if (num_bytes <= 8)
+  {
+    transform_and_sort_pairs_f.template operator()<uint64_t>(0);
+  }
+}
+
+// overload for key_out_t == std::nullptr_t
+template <typename _ExecutionPolicy, typename key_t, typename key_out_t,
+          typename value_t, typename value_out_t>
+typename ::std::enable_if<::std::is_null_pointer<key_out_t>::value>::type
+sort_pairs_impl(_ExecutionPolicy &&policy, key_t keys_in, key_out_t keys_out,
+                value_t values_in, value_out_t values_out, int64_t n,
+                bool descending, int begin_bit, int end_bit) {
+  // create temporary keys_out to discard, memory footprint could be improved by
+  // a specialized iterator with a single
+  // unchanging dummy key_t element
+  using key_t_value_t = typename std::iterator_traits<key_t>::value_type;
+  auto temp_keys_out = sycl::malloc_device<key_t_value_t>(n, policy.queue());
+  internal::sort_pairs_impl(std::forward<_ExecutionPolicy>(policy), keys_in,
+                            temp_keys_out, values_in, values_out, n, descending,
+                            begin_bit, end_bit);
+  sycl::free(temp_keys_out, policy.queue());
+}
+// DPCT_LABEL_END
+
+} // end namespace internal
+
+// DPCT_LABEL_BEGIN|sort_pairs|dpct
+// DPCT_DEPENDENCY_BEGIN
+// DplExtrasAlgorithm|sort_pairs_impl
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+template <typename _ExecutionPolicy, typename key_t, typename key_out_t,
+          typename value_t, typename value_out_t>
+void sort_pairs(_ExecutionPolicy &&policy, key_t keys_in, key_out_t keys_out,
+                value_t values_in, value_out_t values_out, int64_t n,
+                bool descending, int begin_bit, int end_bit) {
```

Comment:
can begin_bit and end_bit have default value? like:  begin_bit=0,  end_bit=sizeof(key_t)*8;

Reply:
Sure, I don't have a problem with that.  I'll put it in in the sandbox and add some tests, then get it in this PR.

Do you think that descending should be defaulted as well (I would assume to false)?

Reply:
Done.  We accept iterator types rather than just data pointers, so we use:
`sizeof(typename std::iterator_traits<key_t>::value_type) *8)` instead of `sizeof(key_t)*8`

This was originally in there, but I accidentally dropped the defaults when removing some forward declarations where are not necessary for now.